src/infra/core/src/ingest/push_ingest_service_impl.rs

```diff
@@ -0,0 +1,368 @@
+                // them directly.
+                cfg_if::cfg_if! {
+                    if #[cfg(unix)] {
+                        use std::os::unix::fs::FileTypeExt;
+                        let ft = p.metadata().int_err()?.file_type();
+                        if ft.is_fifo() || ft.is_char_device() || ft.is_block_device() {
+                            let temp_path = args.operation_dir.join(ingest_common::get_random_cache_key("read-"));
+                            tracing::info!(
+                                from_path = %p.display(),
+                                to_path = %temp_path.display(),
+                                "Detected a special file type - copying into temporary path first",
+                            );
+                            Self::copy_special_file(&p, &temp_path).await?;
+                            Ok(temp_path)
+                        } else {
+                            Ok(p)
+                        }
+                    } else {
+                        Ok(p)
+                    }
+                }
+            }
+            _ => Err(format!("Unsupported source: {}", args.url).int_err()),
+        }?;
+
+        if input_data_path.metadata().int_err()?.len() == 0 {
+            tracing::info!(path = ?input_data_path, "Early return due to an empty file");
+            return Ok(None);
+        }
+
+        let conf = if let Some(media_type) = &args.media_type {
+            let conf = self
+                .data_format_registry
+                .get_compatible_read_config(args.polling_source.read.clone(), media_type)?;
+
+            tracing::debug!(
+                ?conf,
+                "Proceeding with best-effort compatibility read configuration"
+            );
+            conf
+        } else {
+            args.polling_source.read.clone()
+        };
+
+        let temp_path = args.operation_dir.join("reader.tmp");
+        let reader = self
+            .data_format_registry
+            .get_reader(args.ctx.clone(), conf, temp_path)
+            .await?;
+
+        let df = reader.read(&input_data_path).await?;
+
+        Ok(Some(df))
+    }
+
+    // TODO: This is needed because tokio::fs::copy refuses to work with special
+    // files
+    async fn copy_special_file(
+        source_path: &Path,
+        target_path: &Path,
+    ) -> Result<(), InternalError> {
+        let mut source = std::fs::File::open(source_path).int_err()?;
+        let mut target = std::fs::File::create(target_path).int_err()?;
+
+        tokio::task::spawn_blocking(move || -> Result<(), std::io::Error> {
+            use std::io::{Read, Write};
+            let mut buf = [0u8; 2048];
+            loop {
+                let read = source.read(&mut buf)?;
+                if read == 0 {
+                    break;
+                }
+                target.write_all(&buf[..read])?;
+            }
+            Ok(())
+        })
```

Comment:
A code like this is repeating from time to time in many places all over the project. Can we think of reusing these adaptations universally?

Reply:
I managed to replace the low level code with `std::io::copy` and `tokio::io::copy`.